hles/transformation/src/main/scala/org/broadinstitute/monster/dap/CaseClassInspector.scala

```diff
@@ -0,0 +1,41 @@
+package org.broadinstitute.monster.dap
+
+import scala.reflect.runtime.universe.{Symbol, Type, TypeTag, typeOf, termNames}
+
+object CaseClassInspector {
+
+  def snakeCaseHeaderRow[T: TypeTag](delimiter: String = "\t"): String =
+    snakeCaseHeaderList[T].mkString(delimiter)
+
+  def snakeCaseHeaderList[T: TypeTag]: Seq[String] = snakeCaseHeaderList(typeOf[T])
+
+  private def snakeCaseHeaderList(t: Type) =
+    flattenParamsList(t).map(_.name.toString).map(snakify(_))
+
+  private def flattenParamsList(t: Type): Seq[Symbol] = {
+    caseClassParamsList(t).flatMap(expandParamsList(_))
+  }
+
+  private def caseClassParamsList(t: Type): Seq[Symbol] =
+    t.decl(termNames.CONSTRUCTOR).asMethod.paramLists(0)
+
+  private def expandParamsList(field: Symbol): Seq[Symbol] = {
+    getNestedSymbols(field.info).getOrElse(Seq(field))
+  }
+
+  // courtesy of https://github.com/lift/framework/blob/master/core/util/src/main/scala/net/liftweb/util/StringHelpers.scala
+  private def snakify(name: String) =
+    name
+      .replaceAll("([A-Z]+)([A-Z][a-z])", "$1_$2")
+      .replaceAll("([a-z\\d])([A-Z])", "$1_$2")
+      .replaceAll("([a-z])(\\d)", "$1_$2")
+      .toLowerCase
+
+  private def getNestedSymbols(targetType: Type): Option[Seq[Symbol]] = {
+    targetType match {
+      case optional if optional <:< typeOf[Option[Any]] => getNestedSymbols(optional.typeArgs(0))
```

Comment:
❓ What's the `<:<` operator doing here?

Reply:
this is saying "if this class can be cast to `Option[Any]`", i.e. "is it an Option[X] type"

Reply:
since you can't do pattern matching on types, to my knowledge

Reply:
You should be able to pattern match on a type: https://docs.scala-lang.org/tour/pattern-matching.html (sorry the page doesn't anchor link, if you search for "Matching on type only" it should jump to the right section)

Reply:
@raaidbroad sadly not, that only lets you match if a value is an instance of the given type. these are actual type objects, so this is essentially matching on value.